### ISSUE TITLE
Update JT12

### ISSUE THREAD

**Issue by jotego** [MiSTer-devel/Genesis_MiSTer#209]
A recent change in [JT12](https://github.com/jotego/jt12/commit/0b5cf576ca5c393db5c8c4df38668a7e6a5ddf7e), made to fix a different system, fixed [this demo](https://www.pouet.net/prod.php?which=81513) in Megadrive (tested by Gyurco on MiST).

Please consider updating the files from JT12 in this repository.

**Comment by birdybro** [comment]
Agreed, the current module is about 3 years old I believe. I've gotten very close to successfully updating them in the past, but was stopped by the voice samples in Aladdin running at 4x speed (and other sounds made like it). I have been meaning to try again as I've gained a little more experience with this stuff... I think if it were to be updated, the way the ladder effect is designed at other places would have to be redone a little bit since JT12 handles this internally now (from what I could see). I think the ladder effect was first implemented in this core and then later added to JT12 if I'm not mistaken, so that kinda dates how old JT12 is here.

Also, I think the new version of JT89 should be evaluated for update as there are various PSG bugs present in the core. Some of the bugs might be due to VDP timing (as the Sega 32X core likely revealed) and others might be due to things that the newer versions of JT89 has fixed (like the Striker whistle bug).

**Comment by sorgelig** [comment]
Ok. i will update it.
There is always something gets broken in jt12 with update, that's why i didn't hurry to update it.

**Comment by sorgelig** [comment]
i've updated jt12. It seems channel mixing balance is a little different than older version. Music has different accents. Also older version had special tweak to reproduce YM3438 mixing. Since new jt12 is quite different in that part, i cannot port the changes.
I've pushed the changes, so anyone can test and tell if anything got fixed or broken.

**Comment by jotego** [comment]
Maybe just copying the jt12_timer file will be enough to get [that demo](https://www.pouet.net/prod.php?which=81513) working. That way you can keep the tweaks in this repo's version.
Still, it's worth checking a few games before publishing even if you only copy one file.

**Comment by birdybro** [comment]
Before
![DA__ALL_A-MD1UJVA6-M192_vs_MiSTer_Stable_20220422_0000](https://user-images.githubusercontent.com/16388068/164874682-b557d403-67f2-4e11-8fdb-88fb20abf8b1.png)

After
![DA__ALL_A-MD1UJVA6-M192_vs_MiSTer_unstable_20220422_0000](https://user-images.githubusercontent.com/16388068/164874687-5aeadce0-f2ba-4c1d-87f7-4dad188d1f0e.png)

The FM channel is now missing frequencies when compared to a sample from @ArtemioUrbina. Here's the samples I compared:

[BirdyBro-04-02-2022.zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/8546112/BirdyBro-04-02-2022.zip)

MiSTer_Stable is the current release, MiSTer_Unstable is with the latest changes with the updated jt12. So this should be addressed before a release is compiled as it made the FM channel much different from original hardware, measurably. My HDMI Capture Card isn't great, so hopefully someone else can check their results too and compare.

**Comment by sorgelig** [comment]
updated jt89 as well

**Comment by birdybro** [comment]
Here's the updated MDFourier plot:

![DA__ALL_A-MD1UJVA6-M192_vs_MiSTer_unstabler_jt89_20_0000](https://user-images.githubusercontent.com/16388068/164930315-55fbb355-d526-4f9d-9e0d-5f869b8e97ec.png)

**Comment by SegaSnatcher** [comment]
> Here's the updated MDFourier plot:

Is the FM still measuring too low in that plot with latest updates?

**Comment by birdybro** [comment]
Yes, it's not so much that FM is measuring too low, it's that PSG is too high relative to FM. Since all Genesis have very different sound, I'm only comparing Model 1 to Model 1, but the core may show all sound at different levels dependent upon frequency across different revisions of Model 1's.

**Comment by sorgelig** [comment]
older jt12 version had very low volume, so it was amplified externally. New version it seems already amplified internally. I cannot tell for sure but new version sounds quieter it seems.

**Comment by birdybro** [comment]
Making PSG quieter by 1dB should return it to the same as it was.

**Comment by sorgelig** [comment]
is it 1/4 times?

**Comment by birdybro** [comment]
https://github.com/MiSTer-devel/Genesis_MiSTer/blob/6d671f2951e44b638aa3917ab458ad89fc68d248/rtl/system.sv#L1432 - I think this is the relevant line to change, probably either change it to 4 or 6 here, whichever would reduce it in one direction. This is one of the lines that Kitrinx added when she adjusted the volume in this commit --> https://github.com/MiSTer-devel/Genesis_MiSTer/commit/195d16cad5e1a708a27fc0323531b2e4c7482f0d#diff-8e1df17dc376cf4f58245460d95b94fee9a3318960cacb5d8673abde40994c88R422

**Comment by sorgelig** [comment]
while decreasing volume (on PSG) should be more safe, overall volume of core may be lowered. 
But at the same time i'm not sure if volume of jt12 can reach max or not (as before).

**Comment by birdybro** [comment]
Shadow of the Beast is one of the loudest games on the system, it's a good test of what the loudest sounds are, I'll check that real quick...

**Comment by sorgelig** [comment]
ok, i'm compiling

**Comment by birdybro** [comment]
Here's the comparison with Shadow of the Beast. There is this subtle "whistling" distortion in the background at times, that is actually how original hardware often behaved, so I'm convinced this is a good update with increased accuracy, despite the potential for some bugs that may come up. I'm going to do real hardware comparisons, have to setup my equipment again which will take a bit...

[Before and After.zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/8548643/Before.and.After.zip)

**Comment by sorgelig** [comment]
try this build:
[Genesis.zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/8548652/Genesis.zip)

**Comment by birdybro** [comment]
Before (meaning with the jt89 update on the main branch) compared with current release 20220326:
![before](https://user-images.githubusercontent.com/16388068/164947210-a35a5b56-f80e-43be-88ba-475fd55b9a91.png)

After:
![after](https://user-images.githubusercontent.com/16388068/164947232-a1b731d0-266c-4964-b66d-866b61c5b2d0.png)

Looks like it shifted from +1.5dB to -1dB in a relative comparison.

**Comment by sorgelig** [comment]
it seems graphs are relative. It shows jt89 volume shifted while i didn't touch it. If i understood right you used build from latest sources i've updated and the rbf i gave above.
I've changed volume of jt12 while your graphs show change of jt89. confusing..

**Comment by birdybro** [comment]
Yes it's relative, the center line is 0, so when you raised jt12 this brought jt12 closer to jt89. JT89 is currently too loud relative to FM when compared to the previous release, so the goal should probably be to reduce jt89's volume.

**Comment by sorgelig** [comment]
try this:
[Genesis2.zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/8548708/Genesis2.zip)

**Comment by sorgelig** [comment]
so, it was actually not enough gain for jt12? then new build is wrong. 
I need to increase gain for FM then

**Comment by birdybro** [comment]
> try this: [Genesis2.zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/8548708/Genesis2.zip)

That did the trick! :)

![DA__ALL_A-MD1UJVA6-M192_vs_sorg-test_0000](https://user-images.githubusercontent.com/16388068/164948284-137981a4-1fe2-4a80-93f7-c7d88f32ff0d.png)

**Comment by sorgelig** [comment]
oh.. so consider it as fixed?

**Comment by birdybro** [comment]
I'd consider the volume balance fixed now!

Now it sounds like the ladder effect of the ym2612 (which is not present in the ym3438) is not included, which I saw in the changes. Currently the core sounds more like a ym3438 than a ym2612, despite jt12 defaulting to ym2612 in the comments. I think after that is reimplemented a new release could be pushed out. successfully.

**Comment by SegaSnatcher** [comment]
> I'd consider the volume balance fixed now! :)

Just looking for some clarification.  Are the amplitude percentages on the bottom left not relevant to this situation of volume balance?  The reason I ask is because I see FM use to show 99.30% before these new changes and now its down to 79.21%.  Thanks

**Comment by birdybro** [comment]
Most of that is due to the difference of 48kHz and 96kHz. I recorded at different sampling rates. There are also differences due to the ladder effect not being present. Also because there were changes to 240p suite since the reference version, I assume. The main thing I'm looking at is the overall volume comparison.

**Comment by birdybro** [comment]
https://github.com/MiSTer-devel/Genesis_MiSTer/pull/210 - This pull request fixes the audio balance and restores the ladder effect (the ym2612/ym3438 option, ym2612 has the ladder effect, ym3438 does not).

**Comment by birdybro** [comment]
It looks like @jotego will need to add the ladder effect into his jt12 module. My attempts to add that will change the logic of his core and will change fixes he made in the newer versions. I'm unfortunately not skilled enough to do this correctly. The core shouldn't be updated until the JT12 repo is updated with the ladder effect, as the simulation of the ym2612 without it is not accurate.

**Comment by birdybro** [comment]
the update to jt12 caused a regression, the Bimini Run fast sound bug has returned --> https://github.com/MiSTer-devel/Genesis_MiSTer/issues/26

**Comment by thehughhefner** [comment]
My suggestion is to compare the changed files in that fixed #26 with what we currently have.

**Comment by birdybro** [comment]
I referenced the commit which fixed the bug in my comment above.

**Comment by sorgelig** [comment]
If you think you can add ladder to a new jt12, then please open PR in jt12 repo. When it will be merged there, then it's not a problem to update genesis repo after that.

**Comment by birdybro** [comment]
I will look into continuing to learn to add an internal switch for the behavior that would work. Until then, there is more info on the Bimini Run regression as a result of this update: 

https://github.com/MiSTer-devel/Genesis_MiSTer/issues/26#issuecomment-1117716973

The regression for Bimini Run also occurs in SiDi (meaning latest MiST with updated jt12/jt89 has the same regression most likely).

I also tested other games that stream audio (sort of like a PWM method) to the ym2612 and didn't notice any problems. So this game must rely on specific timer behavior that has regressed since April 2019

**Comment by jotego** [comment]
I was wondering if once I add the ladder effect to the original repository, could we use a **git submodule** here instead of copying the files?

I know @sorgelig is worried about loosing the original files. But, there is an easy way around. **Make a fork of JT12 as a repository in mister-devel. Then use git submodule to add that fork to this repository**. Now you have a permanent copy of the files, but you can still PR to the original repository and we can keep track of everything because the file history is not broken.

This approach has the best of both worlds.

I would suggest doing this for other common modules too.

**Comment by birdybro** [comment]
Just my two cents, even though I have no impact on the decision.

I get where you are coming from regarding submodule use because it helps track regressions back into the upstream, and I know lots of devs in various projects use them for good reasons, but there are some specific downsides that would impact us as well:

1. It can effect the portability of the full repo. If we migrate to GitLab for some random reason, then github specific ways of handling submodules is broken and would have to be fixed.
2. Git doesn't download submodules by default, you have to go out of your way to tell it to, and most devs are not used to that workflow consideration. Having to do `git submodule update` and `git clone --recursive` instead of the regular workflow, is extra work.
3. If one repo uses submodules then it's so inconsistent of an experience for the devs that it woudl be a slippery slope toward making all repos use submodules. For the sake of consistently, using submodules in only some cases can be a hindrance.
4. Collaborators don't see updates to submodules without running `git submodule update`, adding an extra step that devs are not used to into the workflow.

The ups and downs of using submodules is described very well here --> https://github.blog/2016-02-01-working-with-submodules/ - it's a tradeoff, you lose some convenience for some added functionality that may see some benefit.

If you implement the ladder effect I will make it a priority to bump updates and test them as they come. I'm watching all activity in your jt12 repo now. I care very deeply about sound on the Genesis core.

**Comment by sorgelig** [comment]
I revisited the idea of submodules several times already and it still gives more inconveniences. For git you always have to issue submodule update command so it's easy to forget it and use older local copy. Inconvenient tracking of changes as history is not included into main repo. Even being linked to submodule repo, it's complicated to see where in the submodule history current link is. And many other not so obvious issues.
So a lot of headache for basically nothing.

**Comment by jotego** [comment]
_git submodules_ do require the user to learn some new skills. However, after having used it for several years and on many projects, I can say that it pays off. Anyway, thanks for having considered it. I may bring the topic up again in 2024.

**Comment by birdybro** [comment]
This seems to be completed and can be closed. I'm testing jt89 being bumped at the moment for regressions btw.
